### PR TITLE
Toolchain: add an experimental CMakeLists.txt for building toolchain

### PR DESCRIPTION
**Pull request by predmond**
This cmake file implements the functionality of BuildIt.sh and BuildQemu.sh.
It is only tested on linux so it likely breaks things that work on other
platforms.

~~Additionally, support for x86_64 isn't implemented but that should be
straight-forward.~~

BuildFuseExt2.sh and BuildPython.sh functionality isn't implemented but those
should also be straight-forward.

To try building the toolchain with cmake use:
```
cd ${SERENITY_ROOT}
cmake -Bbuild/toolchain Toolchain
cmake --build build/toolchain
```

### CODE REVIEW

**Comment by predmond** [comment]
This obviously doesn't have the same functionality as the scripts but I was interested in knowing:
1) is this a desirable change (the CMakeLists.txt is quite a bit simpler than the scripts)
2) does it work on other platforms
3) how might the travis-ci caching fit in?

**Comment by predmond** [comment]
also note: you can use
`cmake --build build/toolchain --target qemu` to build qemu

File: Toolchain/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+set(PREFIX ${CMAKE_CURRENT_SOURCE_DIR}/Local)
+set(TARGET i686-pc-serenity)
+set(SYSROOT ${CMAKE_CURRENT_SOURCE_DIR}/../Root)
+set(PATCHES ${CMAKE_CURRENT_SOURCE_DIR}/Patches)
+
+set(LIBC ${CMAKE_CURRENT_SOURCE_DIR}/../Libraries/LibC)
+set(LIBM ${CMAKE_CURRENT_SOURCE_DIR}/../Libraries/LibM)
+
+ExternalProject_Add(binutils
+    URL      "http://ftp.gnu.org/gnu/binutils/binutils-2.33.1.tar.gz"
```
Comment (bcoles) [inline Toolchain/CMakeLists.txt:20]: Reason not to use https?

```suggestion
    URL      "https://ftp.gnu.org/gnu/binutils/binutils-2.33.1.tar.gz"
```

File: Toolchain/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        --enable-shared
+        --disable-nls
+    BUILD_COMMAND ${MAKE_EXE} -j${MAKE_JOBS}
+    INSTALL_COMMAND ${MAKE_EXE} install
+)
+
+ExternalProject_Add(gcc
+    DEPENDS binutils
+
+    URL      "http://ftp.gnu.org/gnu/gcc/gcc-9.3.0/gcc-9.3.0.tar.gz"
```
Comment (bcoles) [inline Toolchain/CMakeLists.txt:38]: Reason not to use https?

```suggestion
    URL      "https://ftp.gnu.org/gnu/gcc/gcc-9.3.0/gcc-9.3.0.tar.gz"
```

**Comment by lnzero1dev** [comment]
Hi @predmond,

I built a build system generator, that is actually an abstraction of CMake or other build systems. For now, there exists only a cmake generator. But it would be nice to also have other generators in the future.

If you're interested, have a look at: https://github.com/lnzero1dev/meta
The build data is put into JSON files and the meta tool reads them and generate the toolchain / host / target CMakeLists.txt files and you should able to build. 

As I write I upload a patch for the JSON files to match the current serenity master. The files for serenity are in the folder "serenity" :^) If you are interested or have any questions, I would be happy if somebody is testing that out :-)

**Comment by lnzero1dev** [comment]
Hey @predmond, did you had any time to look at the meta generator? If yes, has it worked out?

**Comment by awesomekling** [comment]
I'm not sure I understand what this PR improves(?)

**Comment by awesomekling** [comment]
Closing this due to lack of answers/activity. If this is picked up again, we can just reopen it :)

**Pull request closed without merging**